tokio-timer/src/debounce.rs

```diff
@@ -0,0 +1,344 @@
+//! Debounce streams on the leading or trailing edge or both edges for a certain
+//! amount of time.
+//!
+//! See [`Debounce`] for more details.
+//!
+//! [`Debounce`]: struct.Debounce.html
+
+use {clock, Delay, Error};
+
+use futures::{
+    future::Either,
+    prelude::*,
+};
+use std::{
+    cmp,
+    error::Error as StdError,
+    fmt::{Display, Formatter, Result as FmtResult},
+    time::{Duration, Instant},
+};
+
+/// Debounce streams on the leading or trailing edge or both.
+#[derive(Debug)]
+pub struct Debounce<T: Stream> {
+    delay: Option<Delay>,
+    duration: Duration,
+    edge: Edge,
+    last_item: Option<T::Item>,
+    max_wait: Option<Duration>,
+    max_wait_to: Option<Instant>,
+    stream: T,
+}
+
+/// Builds a debouncing stream.
+#[derive(Debug)]
+pub struct DebounceBuilder<T> {
+    duration: Option<Duration>,
+    edge: Option<Edge>,
+    max_wait: Option<Duration>,
+    stream: T,
+}
+
+/// Either the error of the underlying stream, or an error within tokio's
+/// timing machinery.
+#[derive(Debug)]
+pub struct DebounceError<T>(Either<T, Error>);
```

Comment:
Do we want to unify this with https://github.com/tokio-rs/tokio/pull/736/files#diff-f928893cf2216788f0ccd9f86e2b1aefR21?